serialize: generate hdf_serialize library

dsikich opened the pull request:
The hdf_serialize library is generated if hdf5 is found
on the system, and if it is not then no library is
generated. The library will be used to preserve daos
container properties and user attributes in daos filesystem
when data is moved from DAOS to a regular POSIX filesystem.
The library will also be used to store serialization and
deserialization functions for the future daos serialize
and daos deserialize utilities.

Signed-off-by: Danielle Sikich <danielle.sikich@intel.com>

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,34 @@
+    #If hdf5 installed build hdf_serialize library 
+    conf = Configure(denv)
+    if not conf.CheckLibWithHeader('hdf5', 'hdf5.h', 'c'):
+	have_hdf5 = False 
```
(style)  trailing whitespace
(pylint-syntax-error) inconsistent use of tabs and spaces in indentation (<unknown>, line 30)

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+
+/* for user attr dataset */
+struct usr_attr {
+	char* attr_name;
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+};
+
+static int
+cont_serialize_prop_acl(hid_t *file_id, struct daos_prop_entry* entry,
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+
+static int
+cont_serialize_prop_acl(hid_t *file_id, struct daos_prop_entry* entry,
+			const char* prop_str)
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+	}
+
+	/* convert acl to list of strings */
+	acl = (struct daos_acl *)entry->dpe_val_ptr;                         
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+}
+
+static int
+cont_serialize_prop_str(hid_t *file_id, struct daos_prop_entry* entry,
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+
+static int
+cont_serialize_prop_str(hid_t *file_id, struct daos_prop_entry* entry,
+			const char* prop_str)
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+		rc = 1;
+		goto out;
+	}
+	
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+		fprintf(stderr, "failed to create attribute");
+		rc = 1;
+		goto out;
+	}   
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+
+static int
+cont_serialize_prop_uint(hid_t *file_id,
+			 struct daos_prop_entry* entry,
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+	hsize_t		dims[1];
+	struct usr_attr	*attr_data = NULL;
+	int		i;
+	
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+	hid_t	usr_attr_val_vtype;
+
+	fprintf(stderr, "Writing metadata file: %s\n", filename);
+	
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+		}
+
+		status = H5Tclose(usr_attr_name_vtype);
+        	if (status < 0) {
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+
+		status = H5Tclose(usr_attr_name_vtype);
+        	if (status < 0) {
+	       		rc = 1;
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+}
+
+static int
+cont_deserialize_prop_str(hid_t *file_id, struct daos_prop_entry* entry,
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+
+static int
+cont_deserialize_prop_str(hid_t *file_id, struct daos_prop_entry* entry,
+			  const char* prop_str)
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+		goto out;
+	}
+	size_t buf_size = H5Tget_size(attr_dtype);
+	if (buf_size <= 0) {
```
(style)  Missing a blank line after declarations

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+}
+
+static int
+cont_deserialize_prop_uint(hid_t *file_id, struct daos_prop_entry* entry,
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+
+static int
+cont_deserialize_prop_uint(hid_t *file_id, struct daos_prop_entry* entry,
+			   const char* prop_str)
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+}
+
+static int
+cont_deserialize_prop_acl(hid_t *file_id, struct daos_prop_entry* entry,
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+
+static int
+cont_deserialize_prop_acl(hid_t *file_id, struct daos_prop_entry* entry,
+			  const char* prop_str)
```
(style)  "foo* bar" should be "foo *bar"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+		rc = 1;
+		goto out;
+	}
+	rdata = calloc(attr_dims[0], sizeof(char*));
```
(style)  "(foo*)" should be "(foo *)"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+		fprintf(stderr, "failed to convert acl strs");
+		goto out;
+	}
+	entry->dpe_val_ptr = (void*)acl;
```
(style)  "(foo*)" should be "(foo *)"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+}
+
+static int
+cont_deserialize_all_props(hid_t *file_id, daos_prop_t *prop, 
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+{
+
+	int			rc = 0;
+        struct daos_prop_entry *entry;
```
(style)  code indent should use tabs where possible

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+		return ENOMEM;
+	}
+
+	prop->dpp_entries[0].dpe_type = DAOS_PROP_CO_LABEL; 
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+	}
+	entry = &prop->dpp_entries[16];
+	/* read acl as a list of strings in deserialize, then convert back to
+	 * acl for property entry 
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+	entry = &prop->dpp_entries[16];
+	/* read acl as a list of strings in deserialize, then convert back to
+	 * acl for property entry 
+ 	 */
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+
+int
+deserialize_daos_cont_prop_metadata(char *filename, daos_prop_t *props,
+			  	    uint64_t *cont_type)
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1033 @@
+
+static int
+cont_deserialize_usr_attrs(hid_t *file_id, uint64_t *_num_attrs,
+		       	   char ***_names, void ***_buffers, size_t **_sizes)
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5741/2/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

Note: Error annotation limited to the first 31 errors.  Remaining unannotated errors:

[src/client/serialize/hdf_serialize.c:929](https://github.com/daos-stack/daos/blob/94326e62f0b3273c81f65f2169e1e96b0cc18fe7/src/client/serialize/hdf_serialize.c#L929):
(style)  spaces required around that '=' (ctx:VxW)

[src/client/serialize/hdf_serialize.c:966](https://github.com/daos-stack/daos/blob/94326e62f0b3273c81f65f2169e1e96b0cc18fe7/src/client/serialize/hdf_serialize.c#L966):
(style)  "(foo*)" should be "(foo *)"

[src/client/serialize/hdf_serialize.c:972](https://github.com/daos-stack/daos/blob/94326e62f0b3273c81f65f2169e1e96b0cc18fe7/src/client/serialize/hdf_serialize.c#L972):
(style)  "(foo*)" should be "(foo *)"

[src/client/serialize/hdf_serialize.c:993](https://github.com/daos-stack/daos/blob/94326e62f0b3273c81f65f2169e1e96b0cc18fe7/src/client/serialize/hdf_serialize.c#L993):
(style)  spaces required around that '=' (ctx:VxW)

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+{
+
+	int			rc = 0;
+        struct daos_prop_entry *entry;
```
(style)  code indent should use tabs where possible

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+	int		i;
+
+	/* Read the user attributes */
+	dset= H5Dopen(*file_id, "User Attributes", H5P_DEFAULT);
```
(style)  spaces required around that '=' (ctx:VxW)

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+		rc = 1;
+		goto out;
+	}
+	names = calloc(num_attrs, sizeof(char*));
```
(style)  "(foo*)" should be "(foo *)"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+		rc = 1;
+		goto out;
+	}
+	buffers = calloc(num_attrs, sizeof(void*));
```
(style)  "(foo*)" should be "(foo *)"

daosbuild1 commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+	*_num_attrs = num_attrs;
+	*_names = names;
+	*_buffers = buffers;
+	*_sizes= sizes;
```
(style)  spaces required around that '=' (ctx:VxW)

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,33 @@
+    #If hdf5 installed build hdf_serialize library
+    conf = Configure(denv)
+    if not conf.CheckLibWithHeader('hdf5', 'hdf5.h', 'c'):
+	have_hdf5 = False
```
(pylint-syntax-error) inconsistent use of tabs and spaces in indentation (<unknown>, line 30)

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5741/3/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,32 @@
+    #If hdf5 installed build hdf_serialize library
+    conf = Configure(denv)
+    if not conf.CheckLibWithHeader('hdf5', 'hdf5.h', 'c'):
+	have_hdf5 = False
```
(pylint-syntax-error) inconsistent use of tabs and spaces in indentation (<unknown>, line 30)

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5741/4/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,32 @@
+    src = ['hdf_serialize.c']
+    if have_hdf5 is True:
+        print("Building with hdf5 bindings.")
+    	hdf_serialize = daos_build.library(denv, 'hdf_serialize', src, LIBS=libraries)
```
(pylint-syntax-error) inconsistent use of tabs and spaces in indentation (<unknown>, line 38)

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5741/5/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

brianjmurrell commented:
You need to remove the skip-{build,test} from your commit message to confirm that your changes do build.

dsikich commented:
> You need to remove the skip-{build,test} from your commit message to confirm that your changes do build.
> 
> You will want to add a "BuildRequires: hdf5-devel" (or somesuch) to get the hd5f headers/libs that you want installed to build the new library. You will also want to add the same package(s) to the utils/docker/Dockerfiles used in the Build stages of the Jenkinsfile.

@brianjmurrell would adding hdf5-devel add hdf5 as a dependency to DAOS though? I just want to make sure because we do not want to require hdf5 to build DAOS.

brianjmurrell commented:
> @brianjmurrell would adding hdf5-devel add hdf5 as a dependency to DAOS though? I just want to make sure because we do not want to require hdf5 to build DAOS.

Yes.  *foo*-devel always requires *foo*.  Dependencies are transitory and that should be taken advantage of.

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,31 @@
+    src = ['hdf_serialize.c']
+    if have_hdf5 is True:
+        print("Building with hdf5 bindings.")
+    	hdf_serialize = daos_build.library(denv, 'hdf_serialize', src, LIBS=libraries)
```
(pylint-syntax-error) inconsistent use of tabs and spaces in indentation (<unknown>, line 38)

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5741/6/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/6/execution/node/85/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/6/execution/node/366/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/6/execution/node/419/log

daosbuild1 commented:
Test stage Build on CentOS 7 debug completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/6/execution/node/437/log

daosbuild1 commented:
Test stage Build on CentOS 7 release completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/6/execution/node/438/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/6/execution/node/494/log

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,31 @@
+    if have_hdf5 is True:
+        print("Building with hdf5 bindings.")
+        hdf_serialize = daos_build.library(denv, 'hdf_serialize', src, LIBS=libraries)
+    	denv.Install('$PREFIX/lib64/', hdf_serialize)
```
(pylint-syntax-error) inconsistent use of tabs and spaces in indentation (<unknown>, line 39)

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5741/7/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/7/execution/node/99/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/7/execution/node/361/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/7/execution/node/432/log

daosbuild1 commented:
Test stage Build on CentOS 7 debug completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/7/execution/node/435/log

daosbuild1 commented:
Test stage Build on CentOS 7 release completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/7/execution/node/436/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/7/execution/node/438/log

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,31 @@
+    src = ['hdf_serialize.c']
+    if have_hdf5 is True:
+        print("Building with hdf5 bindings.")
+        hdf_serialize = daos_build.library(denv, 'hdf_serialize', src, LIBS=libraries)
```
(pylint-line-too-long) Line too long (86/80)

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,31 @@
+
+    have_hdf5 = True
+
+    """Do hdf_serialize configure checks"""
```
(pylint-pointless-string-statement) String statement has no effect

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,31 @@
+    conf = Configure(denv)
+    if not conf.CheckLibWithHeader('hdf5', 'hdf5.h', 'c'):
+        have_hdf5 = False
+    denv = conf.Finish()
```
(pylint-assignment-from-no-return) Assigning result of a function call, where the function has no return

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,31 @@
+"""Build FUSE client"""
+import daos_build
+
+def scons():
```
(pylint-inconsistent-return-statements) Either all return statements in a function should return an expression, or none of them should.

daosbuild1 replied:
(pylint-inconsistent-return-statements) Either all return statements in a function should return an expression, or none of them should.

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5741/8/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/8/execution/node/99/log

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5741/9/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/9/execution/node/99/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

dsikich commented:
> > @brianjmurrell would adding hdf5-devel add hdf5 as a dependency to DAOS though? I just want to make sure because we do not want to require hdf5 to build DAOS.
> 
> Yes. _foo_-devel always requires _foo_. Dependencies are transitory and that should be taken advantage of.

@brianjmurrell ok, I guess at this point I am failing to see why we need to add. hdf5-devel to BuildRequires because we do not want to require hdf5 for the build. We just want to generate this library if hdf5 is already present on the system.

brianjmurrell commented:
> @brianjmurrell ok, I guess at this point I am failing to see why we need to add hdf5-devel to BuildRequires because we do not want to require hdf5 for the build. We just want to generate this library if hdf5 is already present on the system.

Because the _hdf5_ package provides the run-time shared libraries and _hdf5-devel_ provides the development libraries and headers.

dsikich commented:
> > @brianjmurrell ok, I guess at this point I am failing to see why we need to add hdf5-devel to BuildRequires because we do not want to require hdf5 for the build. We just want to generate this library if hdf5 is already present on the system.
> 
> Because the _hdf5_ package provides the run-time shared libraries and _hdf5-devel_ provides the development libraries and headers.

@brianjmurrell yes, thanks for clarification, but maybe I haven't been clear about what I am confused about. I am more concerned about whether or not adding them makes them *required* or *optional* for building DAOS.

daltonbohning commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+	attr_dtype = H5Tcopy(H5T_C_S1);
+	if (attr_dtype < 0) {
+		fprintf(stderr, "failed to create acl type");
+		rc = 1;
+		goto out;
```
Could we use D_GOTO in this file?
```suggestion
		D_GOTO(out, rc = 1)
```

daltonbohning commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+	acl = (struct daos_acl *)entry->dpe_val_ptr;
+	rc = daos_acl_to_strs(acl, &acl_strs, &len_acl);
+	if (rc != 0) {
+		fprintf(stderr, "failed to convert acl to strs");
```
Could we use D_ERROR in this file?
At the least, we should probably add a '\n' at the end of each message.
```suggestion
		D_ERROR("failed to convert acl to strs "DF_RC"\n", DP_RC(rc));
```

mchaarawi replied:
agreed. this is internal DAOS code now. we should use the proper error macros and memory allocation / frees

daltonbohning commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+
+	/* convert acl to list of strings */
+	acl = (struct daos_acl *)entry->dpe_val_ptr;
+	rc = daos_acl_to_strs(acl, &acl_strs, &len_acl);
```
We need to make sure we free `acl_strs` when we're done:
https://github.com/daos-stack/daos/blob/41a9950b0d30acf90a4278efcfa9a0690cd06f6e/src/include/daos_security.h#L574-L575

daltonbohning commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+
+	entry = &prop_query->dpp_entries[0];
+	rc = cont_serialize_prop_str(file_id, entry, "DAOS_PROP_CO_LABEL");
+	if (rc != 0) {
+		goto out;
+	}
```
Optional: Brackets are okay, but DAOS style does allow NOT putting.

daltonbohning commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+	/* Create the user attribute data space */
+	dims[0] = num_attrs;
+	dspace = H5Screate_simple(1, dims, NULL);
+	if (dspace < 0) {
+		fprintf(stderr, "failed to create user attr dspace");
+		rc = 1;
+		goto out;
+	}
```
In general, for this API, I'm not sure what we want the return codes to be. In mpifileutils we use generic 1/-1 for a lot of things, but in DAOS the standard is to use the -DER codes. We might need to propagate errors up from helper functions and return the corresponding -DER codes, or MISC/generic for HDF5 errors.
Probably best to double check with someone who knows better.

daltonbohning commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+	}
+
+	/* Allocate space for all attributes */
+	attr_data = calloc(num_attrs, sizeof(struct usr_attr));
```
Can we use D_ALLOC in this file? (D_ALLOC works like calloc)

daltonbohning commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+	}
+
+out:
+	free(attr_data);
```
Can we use D_FREE in this file?

daltonbohning commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+	hid_t	usr_attr_name_vtype;
+	hid_t	usr_attr_val_vtype;
+
+	fprintf(stderr, "Writing metadata file: %s\n", filename);
```
I'm not sure how messages like this should be printed for this API. We might need to use the standard INFO, DEBUG, ERROR masks.

daltonbohning commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+	int		i;
+
+	/* Read the user attributes */
+	dset = H5Dopen(*file_id, "User Attributes", H5P_DEFAULT);
```
"User Attributes" should probably be captured in a #DEFINE at the top of this file

daltonbohning reviewed (commented):
Should this have a corresponding header file? (hdf_serialize.h)

johannlombardi commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+int
+deserialize_daos_cont_attrs_metadata(char *filename, uint64_t *num_attrs,
+				     char ***names, void ***buffers,
+				     size_t **sizes)
```
That's a lot of wildcards ...

johannlombardi commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+
+
+int
+deserialize_daos_cont_attrs_metadata(char *filename, uint64_t *num_attrs,
```
We should probably have some naming convention for symbols exported by this library. Like daos_cont_serialize_md/attrs/...

johannlombardi commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,29 @@
+    if have_hdf5 is True:
+        print("Building with hdf5 bindings.")
+        hdf_serialize = daos_build.library(denv, 'hdf_serialize', src,
+                                           LIBS=libraries)
```
The library name should help to recognize that this is daos-related to avoid confusion.

brianjmurrell commented:
> @brianjmurrell yes, thanks for clarification, but maybe I haven't been clear about what I am confused about. I am more concerned about whether or not adding them makes them _required_ or _optional_ for building DAOS.

Ah, yes.  I see I misunderstood your question.

So while I agree completely with the source build allowing the hdf5 library to be absent and not build the optional functionality from it -- allowing the user who builds from source to choose if they want that functionality -- it's standard practice for binary packaging to include as many features as possible.  This is because we only produce one binary package and leaving features out makes them unsuitable for some class of users, while adding them in makes the binary packages more widely accepted and usable.

If you are concerned with the users of the binary package having to have hdf5 on their systems, you can package your optional feature (library) into a separate _suib-package_ and then people who want to avoid the hdf5 run-time dependency can simply opt not to install that sub-package.

mchaarawi commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,29 @@
+"""Build FUSE client"""
```
fuse?

mchaarawi commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,29 @@
+    src = ['hdf_serialize.c']
+    if have_hdf5 is True:
+        print("Building with hdf5 bindings.")
+        hdf_serialize = daos_build.library(denv, 'hdf_serialize', src,
```
i don't think hdf_serializer is a suitable name for this.
hdf is very common to many libraries. we are using HDF5 here. this is also a DAOS tool not an HDF5 one.
serialize is also very generic. we will support serializing just not the md part?
so we should be calling this something like daos_md_serialize or something like that?

dsikich replied:
@mchaarawi yes this will support more than just the attribute part. It will also contain the hdf5 serialization code necessary for the daos serialize/deserialize tools. What about naming it just daos_serialize?

mchaarawi replied:
im fine with whatever as long as you don't call it hdf

mchaarawi commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,29 @@
+                                           LIBS=libraries)
+        denv.Install('$PREFIX/lib64/', hdf_serialize)
+    else:
+        print("No installed hdf5 detected, not building with hdf5 enabled")
```
i would clear up that message to mention what is really not enabled.

mchaarawi commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+		rc = 1;
+		goto out;
+	}
+out:
```
you leak a lot of HDF5 objects on failure. should make sure those are closed depending where you fail.
this applies to most functions in this file.

mchaarawi commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+};
+
+static int
+cont_serialize_prop_acl(hid_t *file_id, struct daos_prop_entry *entry,
```
why is file_id passed by reference in all those functions?

mchaarawi commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+	int			rc = 0;
+	struct daos_prop_entry	*entry;
+
+	entry = &prop_query->dpp_entries[0];
```
how are you going to guarantee that whenever one adds a new property, they are going to remember that they add the serializing here?
can we do something like a loop to get all container properties and serialize them vs hardcoding each one?

dsikich replied:
@mchaarawi we have a list of the container properties serialized documented in the milestone documentation for serialization. I suppose I didn't realize that the container properties would change very often. I also had the initial thought of serializing each container property as its own attribute so that the name of the attribute was the container property name, and the attribute value was the container property value. I suppose we could serialize the container properties similarly to the way that the user attributes are serialized, which is done using an hdf5 compound datatype for usr attribute name and the usr attribute value, which allows varying lengths and datatypes. I could create a container property compound datatype that contains a container property name and a container property value. Is this the better option?

dsikich replied:
@mchaarawi also, there is this issue now: https://jira.hpdd.intel.com/browse/DAOS-7765 where every POSIX container should have a unique label. So is it best to serialize/deserialize the container label property? Because if someone serializes then deserializes a container and an old one still exists then the container label will not be unique?

mchaarawi replied:
as for the label, yes we should serialize it. if a container with the same label exists in the pool where we are deserializing, we should fail, or provider the option to change the label.

mchaarawi commented on src/client/serialize/hdf_serialize.c:
```diff
@@ -0,0 +1,1034 @@
+
+	attr_dims[0] = 1;
+	attr_dtype = H5Tcopy(H5T_NATIVE_UINT64);
+	status = H5Tset_size(attr_dtype, 8);
```
i don't understand why you are copying the datatype and then setting the size.. 
why can't you just use H5T_NATIVE_UINT64 for the attr datatype?

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/11/execution/node/119/log

daosbuild1 commented on utils/rpms/daos.spec:
```diff
@@ -246,6 +246,15 @@ Requires: %{name}-server%{?_isa} = %{version}-%{release}
 %description firmware
 This is the package needed to manage server storage firmware on DAOS servers.
 
+%package hdf_serialize 
```
(style)  trailing whitespace

daosbuild1 commented on utils/rpms/daos.spec:
```diff
@@ -246,6 +246,15 @@ Requires: %{name}-server%{?_isa} = %{version}-%{release}
+%package hdf_serialize 
+Summary: DAOS serialization library that uses HDF5
+BuildRequires: hdf5-devel
+Requires: hdf5 
```
(style)  trailing whitespace

daosbuild1 commented on utils/rpms/daos.spec:
```diff
@@ -246,6 +246,15 @@ Requires: %{name}-server%{?_isa} = %{version}-%{release}
+BuildRequires: hdf5-devel
+Requires: hdf5 
+
+%description hdf_serialize 
```
(style)  trailing whitespace

daosbuild1 commented on utils/rpms/daos.spec:
```diff
@@ -462,7 +471,13 @@ getent passwd daos_agent >/dev/null || useradd -s /sbin/nologin -r -g daos_agent
 # set daos_firmware to be setuid root in order to perform privileged tasks
 %attr(4750,root,daos_server) %{_bindir}/daos_firmware
 
+%files hdf_serialize 
```
(style)  trailing whitespace

daosbuild1 commented on utils/rpms/daos.spec:
```diff
@@ -462,7 +471,13 @@ getent passwd daos_agent >/dev/null || useradd -s /sbin/nologin -r -g daos_agent
+
 %changelog
+* Thu Jun 03 2021 Danielle M. Sikich <danielle.sikich@intel.com> 1.3.0-17
+- Add DAOS serialization library that requires hdf5 
```
(style)  trailing whitespace

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5741/12/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/12/execution/node/120/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

dsikich commented:
@brianjmurrell I added a sub package for the new library. I know that the name of the library itself needs to be updated, but I would like to make sure that I have added the sub package correctly to the daos.spec file first. Did I add this correctly?

brianjmurrell commented:
> @brianjmurrell I added a sub package for the new library. I know that the name of the library itself needs to be updated, but I would like to make sure that I have added the sub package correctly to the daos.spec file first. Did I add this correctly?

Yes.  Good work.

dsikich commented:
> Should this have a corresponding header file? (hdf_serialize.h)

@daltonbohning I don't think so? It would not be used as all of the library functions have to be looked up with dlsym.

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+#include <stdio.h>
+#include <gurt/common.h>                                                         
```
(style)  trailing whitespace
(style)  line over 80 characters

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+
+	entry = &prop_query->dpp_entries[4];
+	rc = serialize_uint(file_id, entry,
+			         "DAOS_PROP_CO_CSUM_CHUNK_SIZE");
```
(style)  code indent should use tabs where possible

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+
+	entry = &prop_query->dpp_entries[5];
+	rc = serialize_uint(file_id, entry,
+			         "DAOS_PROP_CO_CSUM_SERVER_VERIFY");
```
(style)  code indent should use tabs where possible

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+	D_ALLOC(attr_data, num_attrs * sizeof(struct usr_attr));
+	if (attr_data == NULL) {
+		rc = -DER_MISC;
+		D_ERROR("failed to allocate user attributes "DF_RC"\n", DP_RC(rc));
```
(style)  line over 80 characters

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+
+static int
+deserialize_str(hid_t file_id, struct daos_prop_entry *entry,
+	        const char *prop_str)
```
(style)  code indent should use tabs where possible

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+
+static int
+deserialize_uint(hid_t file_id, struct daos_prop_entry *entry,
+	         const char *prop_str)
```
(style)  code indent should use tabs where possible

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+
+static int
+deserialize_acl(hid_t file_id, struct daos_prop_entry *entry,
+	        const char *prop_str)
```
(style)  code indent should use tabs where possible

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+	if (label == NULL) {
+		return ENOMEM;
+	}
+	label->dpp_entries[0].dpe_type = DAOS_PROP_CO_LABEL; 
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+	*_prop = prop;
+out:
+	/* close container after checking if label exists in pool */
+	if (close_cont) 
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+{
+	int	rc = 0;
+	hid_t	file_id = 0;
+	
```
(style)  trailing whitespace

daosbuild1 commented on src/client/serialize/daos_serialize.c:
```diff
@@ -0,0 +1,1025 @@
+
+static int
+deserialize_attrs(hid_t file_id, uint64_t *_num_attrs,
+	    	  char ***_names, void ***_buffers, size_t **_sizes)
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,28 @@
+import daos_build
```
(pylint-missing-module-docstring) Missing module docstring

daosbuild1 commented on src/client/serialize/SConscript:
```diff
@@ -0,0 +1,28 @@
+import daos_build
+
+def scons():
+    """Execute build"""
```
(pylint-misplaced-future) __future__ import is not the first non docstring statement

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5741/14/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/14/execution/node/87/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/14/execution/node/304/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/14/execution/node/386/log

daosbuild1 commented:
Test stage Build RPM on CentOS 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/14/execution/node/318/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5741/14/execution/node/390/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.
